corehq/util/model_log.py

```diff
@@ -1,45 +1,56 @@
 from enum import Enum
 
-from django.contrib.admin.models import ADDITION, CHANGE, DELETION, LogEntry
-from django.contrib.admin.options import get_content_type_for_model
-from django.utils.encoding import force_text
+from corehq.apps.users.models_sql import HQLogEntry, UpdateDetails
+
+MODEL_PROPERTIES_TO_IGNORE = ['_id', '_rev', 'reporting_metadata', 'password']
 
 
 class ModelAction(Enum):
-    CREATE = ADDITION
-    UPDATE = CHANGE
-    DELETE = DELETION
+    CREATE = HQLogEntry.CREATE
+    UPDATE = HQLogEntry.UPDATE
+    DELETE = HQLogEntry.DELETE
 
 
-def log_model_change(user, model_object, message=None, fields_changed=None, action=ModelAction.UPDATE):
+def log_model_change(domain, user, model_object, message=None, fields_changed=None, action=ModelAction.UPDATE,
+                     can_skip_domain=False, changed_via=None):
     """
-    :param user: User making the change (couch user or django user)
-    :param model_object: The object being changed (must be a Django model)
-    :param message: Message text
-    :param fields_changed: List of model field names that have
+    :param domain: domain where the update was initiated
+    :param user: User making the change (couch user) or SYSTEM_USER_ID
+    :param model_object: The user being changed (couch user)
+    :param message: Optional Message text
+    :param fields_changed: dict of model fields that have changed with current value
     :param action: Action on the model
+    :param can_skip_domain: flag to allow domain less entry
+    :param changed_via: changed via medium i.e API/Web
     """
-    from corehq.apps.users.models import CouchUser
-    if isinstance(user, CouchUser):
-        user = user.get_django_user()
-
-    if message is None and fields_changed is None:
-        raise ValueError("One of 'message' or 'fields_changed' is required.")
+    from corehq.apps.users.util import SYSTEM_USER_ID
 
-    if message is not None and fields_changed is not None:
-        raise ValueError("'message' and 'fields_changed' are mutually exclusive")
+    if not domain and user != SYSTEM_USER_ID and not can_skip_domain:
```

Comment:
We could though that would make it generic.

Its just that, I was trying to ensure we log domain in all places where its possible to do so, so that all relevant log entries are visible in the report when we filter by domain. If there is a generic check like that, its possible domain is not passed at times when available, leading to domain less logs, though there are very few places where a web user is registered, but I didn't want to assume that i knew all. 

And plus there is also https://github.com/dimagi/commcare-hq/blob/99c8a0231dca39f5a32e78c85730e48c69611eab/corehq/apps/sso/backends.py#L112 where domain might not be available